shared/src/test/diff/typegen/TypegenTypedefs.mls

```diff
@@ -71,3 +71,39 @@ let lockA = Lock 20 in let lockB = Lock 30 in (Bank lockA 2000).Better(Bank lock
 //│ // end ts
 
 
+:ts
+class None: {}
+class Some[T]: { value: T }
+type Option[T] = Some[T] | None
+class LinkedList[T]: { head: T; tail: Option[LinkedList[T]] }
+  method Append: T -> LinkedList[T]
+  method Append elem = LinkedList { head = elem; tail = Some { value = this } }
+def None = None {}
+def Some v = Some { value = v }
+//│ Defined class None
+//│ Defined class Some
+//│ Defined type alias Option
+//│ Defined class LinkedList
+//│ Declared LinkedList.Append: ((LinkedList['T] with {tail: None | (Some['b .. 'c] with {value: 'a})}) as 'a) -> 'T -> ((LinkedList['T] with {tail: None | (Some[(LinkedList['T] with {tail: None | (Some['b .. 'c] with {value: 'a})}) as 'a .. 'd] with {value: 'd})}) as 'd)
+//│ Defined LinkedList.Append: ((LinkedList['T] with {tail: None | (Some['b .. 'c] with {value: 'a})}) as 'a) -> ('T & 'd & 'e) -> (LinkedList['T & 'e .. 'T | 'e] with {head: 'd, tail: Some[((LinkedList['T | 'e .. 'T & 'e] with {tail: 'g & (None | (Some['h .. LinkedList['T | 'e .. 'T & 'e] with {tail: 'g}] with {value: 'f}))}) as 'f) & 'i .. ((LinkedList['T & 'e .. 'T | 'e] with {tail: None | (Some[(LinkedList['T | 'e .. 'T & 'e] with {tail: 'g & (None | (Some['h .. LinkedList['T | 'e .. 'T & 'e] with {tail: 'g}] with {value: 'f}))}) as 'f .. 'j] with {value: 'j})}) as 'j) | (LinkedList['T] with {tail: None | (Some[?] with {value: (LinkedList['T] with {tail: None | (Some[(LinkedList['T] with {tail: None | (Some['b .. 'c] with {value: 'a})}) as 'a .. 'k] with {value: 'k})}) as 'k})}) | 'i] with {value: LinkedList['T] with {tail: None | (Some[?] with {value: (LinkedList['T] with {tail: None | (Some[(LinkedList['T] with {tail: None | (Some['b .. 'c] with {value: 'a})}) as 'a .. 'k] with {value: 'k})}) as 'k})}}})
+//│ None: None
+//│ Some: ('value & 'T) -> (Some['T] with {value: 'value})
+//│ // start ts
+//│ export declare class None {
+//│     constructor(fields: {})
+//│ }
+//│ export declare class Some<T> {
+//│     value: T
+//│     constructor(fields: {value: T})
+//│ }
+//│ export type Option = Some<T> | None
+//│ export type a<d, c, T> = Omit<LinkedList<T>, "tail"> & {tail: None | Omit<Some<a<d, c, T>>, "value"> & {value: a<d, c, T>}}
```

Comment:
How come this type is much simpler than the MLscript one? the fact that `Some<a<d, c, T>` is given these type arguments is suspicious. I take it we don't support BoundTypes, but in that case why doesn't it yield an error instead? How do these types get here?

Reply:
I've used this smaller example to reproduce the issue. The issue is with how the return type of `Append` is being resolved.

```scala
class Box[G]: { value: G }
class LinkedList[T]: { head: T; tail: Box[LinkedList[T]] }
  method Append: T -> LinkedList[T]
```

Debug info shows that Append return type is this -

```
Recursive(α22,WithExtension(AppliedType(TypeName(LinkedList),List(α17)),Record(List((tail,WithExtension(AppliedType(TypeName(Box),List(Bounds(Recursive(α21,WithExtension(AppliedType(TypeName(LinkedList),List(α17)),Record(List((tail,WithExtension(AppliedType(TypeName(Box),List(Bounds(α20,α19))),Record(List((value,α21))))))))),α22))),Record(List((value,α22)))))))))
```

Here `α17 = T` and the `α22` and `α21` are two new type variables that are introduced.

Reply:
Can you show me the MLscript pretty-printed type too? It's pretty hard to read this internal representation!

Reply:
Here's from the debug info

```
//│ Defined class LinkedList
//│ Typed as: ((LinkedList[T1'],) -> ((T1',) -> LinkedList[T1']))
//│  where: 
//│ Declared LinkedList.Append: ((LinkedList['T] with {tail: Box['b .. 'c] with {value: 'a}}) as 'a) -> 'T -> ((LinkedList['T] with {tail: Box[(LinkedList['T] with {tail: Box['b .. 'c] with {value: 'a}}) as 'a .. 'd] with {value: 'd}}) as 'd)
//│ Typed as: ((LinkedList[T1'],) -> ((T1',) -> LinkedList[T1']))
```

Reply:
That's not the output I get... Why do you have an extra `Typed as:` line? Please make sure the stuff you put here is accurate to avoid waisting out time.

Also, in this example you can see `Box['b .. 'c]` is applied to a bounds type. Why is your algorithm returning a (wrong) result instead of bailing here? Bounds in invariant position are not expressible in TS.

Reply:
This is my input, the typed as is because of debug mode. I've made sure that latest changes from `mlscript` branch have been merged.

```scala
:NoJS
:ts
:d
class Box[G]: { value: G }
class LinkedList[T]: { head: T; tail: Box[LinkedList[T]] }
  method Append: T -> LinkedList[T]
```

Reply:
> Also, in this example you can see Box['b .. 'c] is applied to a bounds type. Why is your algorithm returning a (wrong) result instead of bailing here? Bounds in invariant position are not expressible in TS.

Aah I think the problem is that type arguments for `AppliedType` are not treated as invariants, they are allowed to retain the polarity they already have. I added this check and it throws the exception and doesn't generates ts for this method.

Reply:
> This is my input

WOW, it's because I didn't have `:ts`, sorry. But this made me realize you re-simplify the entire method types just t show the TS. That should be avoided – type simplification is not cheap (in fact it's probably the asymptotically worse part of the type checker).

Reply:
> Aah I think the problem is that type arguments for AppliedType are not treated as invariants, they are allowed to retain the polarity they already have. I added this check and it throws the exception and doesn't generate ts for this method.

Yes, that's a mistake, but even then I still don't see why the recursive type occurrence arrived in this position! Notice that `'b .. 'c` does not refer to the recursive type.